### PR TITLE
added canvas

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
Being able to draw something opens up many more possibilities.

Today while having dinner I had an idea of how to implement a basic canvas widget and now my prototype is working.

It is storing the image data uncompressed in the widget properties. Maybe someone has a better way to do this but I really want the feature of instant synchronization.

The resolution is exactly 100x100 pixels which gets scaled to the size of the widget. It supports 64 colors that can be defined in the property `colorMap` (only the first 6 are currently defined by default). When drawing it uses the color at index `activeColor`.

To-do list:
- [x]  add presets color maps on UI
    - (@ArnoldSmith86 I'm going to make this kind ugly you can refine it before merging if you want) 
- [x]  change UI organization
- [x]  add composed widget to add widget overlay
- [x]  add color map cycle routine function
- [x]  add border similar to holders
- [x]  change columns preset value to  "c..": "-*)($"
- [x]   merge new compression
- [x]   change default to inc
- [x]   fix grey scale
- [x]   fix default canvas that got broken by the new compression

Extra: (@ArnoldSmith86 )
- [ ]  add export/inport routine function
- [ ]  create import export tool

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3444/pr-test (or any other room on that server)

**Comment by ArnoldSmith86** [comment]
Hmm. It doesn't work on the test server. Checking...

**Comment by ArnoldSmith86** [comment]
Works now.

**Comment by robartsd** [comment]
What about making the resolution of the widget configurable? Should the image data be stored in an array in a single property? Could be nice if each player had a different active color.

**Comment by robartsd** [comment]
I see, you've broken the canvas into 100 regions (10 rows and 10 columns each 10x10 pixels) with each region being a separate property. This should allow multiple players drawing in different regions of the same canvas at the same time without creating conflict.

**Comment by ArnoldSmith86** [comment]
> Should the image data be stored in an array in a single property?

The reason it's split up is because only changing properties are sent to the other clients.

> What about making the resolution of the widget configurable?

Maybe. We can see how far we can push this.

**Comment by ArnoldSmith86** [comment]
> This should allow multiple players drawing in different regions of the same canvas at the same time without creating conflict.

Right now conflicts are on widget level but that will change soon.

**Comment by ArnoldSmith86** [comment]
I feel like 400x400 with 16 colors would be OK to use this in a 1000x1000 "virtual" pixel widget.

For basic scribbles very rudimentary compression will probably make that smaller than it is now for 100x100.

Or do you think this isn't good enough to be useful?

**Comment by robartsd** [comment]
> I feel like 400x400 with 16 colors would be OK to use this in a 1000x1000 "virtual" pixel widget.
> 
> For basic scribbles very rudimentary compression will probably make that smaller than it is now for 100x100.
> 
> Or do you think this isn't good enough to be useful?

400x400 would probably be adequate.

**Comment by 96LawDawg** [comment]
I love the idea and I'm sure many games could take advantage of something like this.  But if I were playing Pictionary, it might be hard to tell a giraffe from a dinosaur.  It's a little bit too pixelated for that.  Could the drawing area by 5x5 pixels?  I don't really understand how you did this, so I don't know if that is possible.  Very cool.

**Comment by ArnoldSmith86** [comment]
I'm starting to really like this.

**Comment by RaphaelAlvez** [comment]
~~~~I tried writing on my phone and it wouldn't works. Later I realized I was going through multiple widgets borders and maybe the widgets are not realizing I'm changing from one to the other.~~~~

~~~~The JSON looks terrible. It is usable for routines which is great but 100 properties may be to much. It is definitely better than one thousand.~~~~

~~~~I'm conflicted with the pixelated look. Don't know how I feel about it yet.~~~~

~~~~I've seen a bunch of SVG versions since the pandemic started. Probably they would be way less useful for automations but I imagine they should also be fairly easy to make compared to what you already made.~~~~

I forgot to press send on my comment yesterday and I was worried when I saw your comment staying you were liking it but then I saw this piece of art:

![Screenshot_20210514-075405.png](https://user-images.githubusercontent.com/73476656/118261375-d14dfb00-b489-11eb-946a-7d1bd8a969a6.png)

My comment from "yesterday" is still somewhat valid so I'm posting both together

**Comment by RaphaelAlvez** [comment]
Ok I'm liking it to. Here are some suggestions:
- changing the resolution is a bit weird going up. Maybe you should fill the space between the points?
- a way to change the brush size (or even the brush shape)
- adding shapes like lines, circles and and rectangles using 2 points as reference
- I understand the need of a compressed files but I can't understand that compression. Maybe go one step further and put it all together in one property that can be used to redraw the whole picture
- your automation is only useful in small images for bigger drawings maybe we should use a matrix

**Comment by ArnoldSmith86** [comment]
> The JSON looks terrible.

Well. Yes. It won't have an ASCII art representation of the image. 😄  It is not optimized for readability but for smallish size and fast synchronization.

> I've seen a bunch of SVG versions

No idea how that would work.

> * Maybe you should fill the space between the points?

If anything I'll add a reset when changing resolution. It's not really meant to be changed while having content in there.

> * a way to change the brush size (or even the brush shape)
> * adding shapes like lines, circles and and rectangles using 2 points as reference

Sure but the way it now it has to be implemented manually. Brush sizes might be easy but if you want bigger brushes, please decrease the resolution.

> * I understand the need of a compressed files but I can't understand that compression. Maybe go one step further and put it all together in one property that can be used to redraw the whole picture

If it's in one property every single pixel change needs to send the whole image.

> * your automation is only useful in small images

True but I'm not trying to replace Photoshop here. All games I have in mind don't need any automation and only very basic scribbles. If you have games that need more, I'd like to hear about them.

**Comment by RaphaelAlvez** [comment]
> > I've seen a bunch of SVG versions
> 
> No idea how that would work.

I **believe** you can set just a line object and you only need to provide hook point to it. (Not really an expert here)

> If anything I'll add a reset when changing resolution. It's not really meant to be changed while having content in there.

That's sounds ok

> Sure but the way it now it has to be implemented manually. Brush sizes might be easy but if you want bigger brushes, please decrease the resolution.

Not the same thing. Try coloring any space in the bigger resolution

> Well. Yes. It won't have an ASCII art representation of the image. 😄  It is not optimized for readability but for smallish size and fast synchronization.
>
> If it's in one property every single pixel change needs to send the whole image.
>
> True but I'm not trying to replace Photoshop here. All games I have in mind don't need any automation and only very basic scribbles. If you have games that need more, I'd like to hear about them.

I'm mostly imagining keeping track of old drawings. I would imagine doing lines , rectangles and circles as the limit in a possible automation need. I do understand that there might be limitations. It did fell really smooth when drawing.

**Comment by ArnoldSmith86** [comment]
> Not the same thing. Try coloring any space in the bigger resolution

Yeah, I'm aware. But I'm not really sure if this is needed. This isn't Photoshop. I'm pretty sure I only know games where you have a time limit for drawing so personally I wouldn't bother with those details while drawing...

But like I said: brush sizes might be easy.

**Comment by robartsd** [comment]
> I'm pretty sure I only know games where you have a time limit for drawing so personally I wouldn't bother with those details while drawing...
> 
> But like I said: brush sizes might be easy.

Sometimes it is worth the time to switch to a pen with a different stroke size, so I'm in favor of adding a brush size option. Other drawing tools (lines & basic shames) could be helpful, but are not analogous to what one would use in a physical drawing game, so I don't think they are needed for VTT.

**Comment by 96LawDawg** [comment]
The color picker on the new UI for the canvas widget does not work.  It just makes the color white.

**Comment by RaphaelAlvez** [comment]
> The color picker on the new UI for the canvas widget does not work. It just makes the color white.

It is working for me. Did you press update to leave the UI? The changes only apply when you press change

**Comment by 96LawDawg** [comment]
Nevermind.  I'm an idiot and was doing it wrong.

I did like the idea of a small widget off to the side that would allow you to pick colors and reset it while in game mode.  Did we do away with that idea?

**Comment by ArnoldSmith86** [comment]
All colors after the first two are black for me on my tablet:

Screenshot_20210522-074938_Privacy_Browser

**Comment by RaphaelAlvez** [comment]
> All colors after the first two are black for me on my tablet:
> 
> Screenshot_20210522-074938_Privacy_Browser

That's because that's an old canvas. Try using the reset colors checkbox.

Your original colors were CSS colors that the html element for some reason doesn't support.

I changed all the colors to the corresponding hex values

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -453,6 +453,37 @@ export class Widget extends StateManaged {
...
+        const execute = async function(widget) {
+          if(widget.get('type') == 'canvas') {
+            if(a.mode == 'setPixel')
+              await widget.setPixel(a.x, a.y, a.colorIndex);
+            else if(a.mode == 'color')
+              await widget.set('activeColor', a.colorIndex === undefined ? null : a.colorIndex);
+            else
+              await widget.reset();
+          }
+        };
```
Comment (RaphaelAlvez) [inline client/js/widgets/widget.js:471]: Maybe that function should be divided into two.  The color side of the function doesn't really have much to do with reset and set and the input is different
There could be more features added related to the color like previous or next color and change to the color map. (mode: set/inc/dec/change, value : number, color:"color string")
Also, I think that there should be an import and export function that puts the whole image into one compressed string. it is kinda impossible to set a good image at the moment.

**Review by RaphaelAlvez — CHANGES_REQUESTED** [review]
I could probably work on those function I just suggesting this for the meeting or discussion here

(not the export feature)

**Review by RaphaelAlvez — COMMENTED** [review]
I believe my part of this PR is ready. @ArnoldSmith86 no need to make the extra stuff in the description stuff before merging this PR

**Comment by RaphaelAlvez** [comment]
preview of default colorMap available on UI

image

http://212.47.248.129:3444/colors

**Comment by bjalder26** [comment]
I'm not saying it needs to be part of this PR, but if you're going to go beyond basic colors, then you might want to consider integrating a version of LawDawg's color picker in the changeRoutine tutorial.

**Comment by RaphaelAlvez** [comment]
> I'm not saying it needs to be part of this PR, but if you're going to go beyond basic colors, then you might want to consider integrating a version of LawDawg's color picker in the changeRoutine tutorial.

that's not really how the canvas work. the colors are not saved as a color value. they are saved as a color index and you can change the whole pallete used in the canvas.
That color picker would need multiple modifications in order to be used.

**Comment by robartsd** [comment]
I haven't looked at the code. Is the palette limited to 10 colors now? Originally a pallet of 64 colors was allowed (I believe using a single base 64 character digit in the encoding), 10 different colors is probably enough, but a standard palate of 64 colors would provide a lot of options without having to add a custom color index to the widget.

**Comment by ArnoldSmith86** [comment]
Someone should do the math for the documentation. I think the limit is around 30 now. The first color uses `"0"` and then they use the next ASCII characters. Somewhere ~30 characters later they start to be used for the compression.

**Comment by robartsd** [comment]
> Someone should do the math for the documentation. I think the limit is around 30 now. The first color uses `"0"` and then they use the next ASCII characters. Somewhere ~30 characters later they start to be used for the compression.

ASCII 35-45 are used to compress color 0 and ASCII 80-116 compress colors 1-15. Valid colors are ASCII 48-79 (32 colors, but only 16 get compressed)

I'd suggest changing the compression so that ASCII 35-45 mean "repeat the previous color" then we could use ASCII 48-116 for color values (69 colors).

**Comment by ArnoldSmith86** [comment]
> or even accept codepoints beyond printable ASCII for more than 79 colors

We could. I felt like it becomes too complicated when considering UTF-8. In theory you could use any Unicode codepoint as a color but then you get multi-byte stuff in UTF-8. Which probably wouldn't be a big problem but I don't think someone wants to make 256 color masterpieces.

> We could also have the assumption that the last color fills the remainder of the region

I like that.

> I'd suggest changing the compression so that ASCII 35-45 mean "repeat the previous color"

I don't think it should just be changed to that without some special cases but it would probably be a good idea for everything above 4 repeating pixels.

Do you have some spare time to implement those changes? Because I really don't have much of it at the moment.

>     * add export/inport routine function

It would actually be kind of easy to have a function that uploads the drawn image to the server and returns an asset URL. And _maybe_ another one that puts the image of an asset URL back into the canvas.

Importing any asset into the canvas would mean that we need a way to reduce the image resolution (the canvas drawing function probably already does that) and more importantly map all the colors from the image to the available canvas colors (https://stackoverflow.com/a/17176082 suggests that it wouldn't need much code).

**Comment by robartsd** [comment]
> > or even accept codepoints beyond printable ASCII for more than 79 colors
> 
> We could. I felt like it becomes too complicated when considering UTF-8. In theory you could use any Unicode codepoint as a color but then you get multi-byte stuff in UTF-8. Which probably wouldn't be a big problem but I don't think someone wants to make 256 color masterpieces.

I think we might just choose to officially support 64-79 indexed colors, but if more work without a change of the code, that's cool too. I don't know much about how Javascript handles unicode.

> > We could also have the assumption that the last color fills the remainder of the region
> 
> I like that.
> 
> Do you have some spare time to implement those changes? Because I really don't have much of it at the moment.

I'll see if I can make time to change the compression as described. It will probably be a bit more computationally expensive than what we have now.

> > I'd suggest changing the compression so that ASCII 35-45 mean "repeat the previous color"
> 
> I don't think it should just be changed to that without some special cases but it would probably be a good idea for everything above 4 repeating pixels.

What special cases are you thinking? I'm thinking the string would have a character representing a color. If the color does not repeat (or only occurs twice in a row), no repeating character would be used. If the color repeats for 3+ pixels the repeating character string up until the next color character would indicate how many pixels to repeat the color for. Not sure if we should keep the base 2 values for the repeating characters or modify the values of the character in some way. Without modification to the values, starting with a blank 500x500 canvas and coloring the last pixel of a region as color 1 would result in the region's property equaling `"0-*)(#1"` though the characters in the middle could be placed in any order without changing the resulting decompressed value (this is also true of the current implementation). Coloring just the first pixel of a blank canvas region would result in the property equaling "10" regardless of resolution (except 10x10 where the property would be "1" since there is only one pixel in the region).

**Comment by ArnoldSmith86** [comment]
Resetting color to greyscale doesn't seem to work.
Changing the dropdown should auto-check the checkbox.
Clicking "Reset colors to:" should check the checkbox (it will if it's a label with the correct `for`).

**Comment by robartsd** [comment]
> > > or even accept codepoints beyond printable ASCII for more than 79 colors
> > 
> > We could. I felt like it becomes too complicated when considering UTF-8. In theory you could use any Unicode codepoint as a color but then you get multi-byte stuff in UTF-8. Which probably wouldn't be a big problem but I don't think someone wants to make 256 color masterpieces.
> 
> I think we might just choose to officially support 64-79 indexed colors, but if more work without a change of the code, that's cool too. I don't know much about how Javascript handles unicode.

I did some research and found that Javascript strings are always treated as if they are composed of 16-bit characters. All code points below U+D800 should be safe to use in arbitrarily manipulated data strings - that would allow for over 55000 colors.

**Comment by ArnoldSmith86** [comment]
Internally it uses UTF-16, yes. But that doesn't mean that everything is valid. But you probably covered that by only going to U+D800.

Keep in mind though that the JSON representation used for transfer and storage is UTF-8. Should be fine but using higher values for compression is weird because they'll use two bytes (or more) instead of one.

File: client/js/widgets/widget.js (hunk 1 above)
Comment (robartsd) [reply client/js/widgets/widget.js:481]: Here is my suggestions for CANVAS modes:
- cycleColor: signed integer indicating how many colors to move (default mode, value defaults to 1)
- setColor: set activeColor index (optionally set the mapped color)
- addColor: append a new color to colorMap
- clear: set all regions to "0"
- setPixel: set the color of a pixel
- getPixel: set a variable to the color index of a pixel
- getRegion: set a variable to the decompressed representation of a region
- setRegion: apply a decompressed representation of a region

**Comment by robartsd** [comment]
> Internally it uses UTF-16, yes. But that doesn't mean that everything is valid. But you probably covered that by only going to U+D800.
> 
> Keep in mind though that the JSON representation used for transfer and storage is UTF-8. Should be fine but using higher values for compression is weird because they'll use two bytes (or more) instead of one.

Yes U+D800 is where the surrogate pairs to access higher Unicode planes start, all code points below that should safely transfer via UTF-8. Abitrary use of U+D800 through U+FFFF could create invalid characters and I'm not sure if UTF-8 would transfer the code points correctly (UTF-8 is capable of encoding more code points in a 32 bit word than UTF-16) My compression uses U+0023 through U+0030 to encode the repeated length. U+0030 through U+D7FF will be available for color values. I was thinking it might be interesting have a fixed 12-bit color range from U+1000 to U+2000 (or an even wider range if we dont care if the value of a particular bit is determined by a single color channel).

**Comment by robartsd** [comment]
I think it would be good to support different aspect ratios. I would propose that the image remain 10x10 regions, but the aspect ratio of the regions be configurable. Of course designers can always use no square widget (with pixel aspect ratio matching widget aspect ratio).

**Comment by robartsd** [comment]
We could make the composed widget easier to resize if the buttons were attached at the top left corner instead of the top right corner.

**Comment by 96LawDawg** [comment]
You cannot select the composed widget in the JSON editor without adding a pixel where you ctrl-click.  This is not the same behavior as other widgets that do not change when ctrl-clicked.

**Comment by RaphaelAlvez** [comment]
> You cannot select the composed widget in the JSON editor without adding a pixel where you ctrl-click.  This is not the same behavior as other widgets that do not change when ctrl-clicked.

Yeah we know. Use F1

**Comment by robartsd** [comment]
> > You cannot select the composed widget in the JSON editor without adding a pixel where you ctrl-click.  This is not the same behavior as other widgets that do not change when ctrl-clicked.
> 
> Yeah we know. Use F1

I don't always use a keyboard with Fn keys.

Also can't move the canvas by dragging (even if the drag occurs on a non-movable child widget).

**Comment by robartsd** [comment]
#512 prevents canvas from handling drags that start on its children as drawing events so a child widget that is not movable can be dragged to move the canvas.

**Comment by robartsd** [comment]
#511 moves the buttons of the composed widget to the left side so that if the canvas is resized, they don't need to be moved to maintain alignment.

**Comment by robartsd** [comment]
Now if movable=false and JSON editor is open, you can use CTRL while dragging to interrupt drawing. If you release CTRL before finishing the drag, then your next mouse move will generate a line from the last drawn point ot the cursor position. However, this seems to have the side effect of messing with the canvas updating itself.

**Comment by RaphaelAlvez** [comment]
> Now if movable=false and JSON editor is open, you can use CTRL while dragging to interrupt drawing. If you release CTRL before finishing the drag, then your next mouse move will generate a line from the last drawn point ot the cursor position. However, this seems to have the side effect of messing with the canvas updating itself.

Yeah I commented this in the last PR. Really not a problem in my opinion. I didn't find any further problems like you discribed

**Comment by robartsd** [comment]
> Yeah I commented this in the last PR. Really not a problem in my opinion. I didn't find any further problems like you discribed

The line drawing part could be considered a feature that we might want implemented without the JSON editor open. The problem with updating itself is a bug.

**Review by 96LawDawg — CHANGES_REQUESTED** [review]
So I've been following this since the beginning.  Very cool tech.  I don't understand any of it.  But it mostly works.  I've been playing with it all along.  In its current form there are only a couple of very minor things that have already been pointed out.  I did just test multiplayer for the first time.  Two people cannot draw on the same canvas at the same time.  That's fine.  Two people can draw on two different canvases at the same time.  Good.

The issue I found came up as a result of trying to test multiplayer on my iPad.  I can't.  I cannot access the test server.  I know about the http(s) issue.  I can access other test servers and the production server from my iPad with no problem.  I typed in the link and I clicked directly from GitHub.  The browser just keeps refreshing the screen and never loads the page.  I have no idea how to look at the "console" equivalent on my iPad (using Safari); it is a slightly older iPad, but not too old.  The test server works on my newer iPhone Safari browser.  Is it possible there is something in the canvas code that makes it incompatible with slightly older iPads?  If so, that is a problem.

**Comment by ArnoldSmith86** [comment]
Does it work now? Seems like your older Safari doesn't know `replaceAll` yet. Fortunately it wasn't really needed by the code (I think).

**Comment by 96LawDawg** [comment]
Yes.  Works now.  I guess I'll do some more testing to make sure whatever you changed didn't break anything.

**Comment by 96LawDawg** [comment]
Final (I think) observations.  In the old-style edit, there is an option to change the palette to pencils, pastels, etc.  I could not get that to work at all.

Also, and this is a very minor thing, when drawing at the same time on two separate canvases I came across a small glitch.  On an empty canvas, when starting to draw first with my desktop, neither my laptop nor my iPad could start drawing until the laptop let go of the mouse.  The other way around (starting with laptop and iPad) was not a problem.  Also, it was not a problem when there was anything on the canvas.  So this would only be a problem in limited simultaneous drawing games where people are drawing at the same time, and even then, only in certain circumstances.

**Comment by robartsd** [comment]
> Also, and this is a very minor thing, when drawing at the same time on two separate canvases I came across a small glitch. On an empty canvas, when starting to draw first with my desktop, neither my laptop nor my iPad could start drawing until the laptop let go of the mouse. The other way around (starting with laptop and iPad) was not a problem. Also, it was not a problem when there was anything on the canvas. So this would only be a problem in limited simultaneous drawing games where people are drawing at the same time, and even then, only in certain circumstances.

So you start with one room with two blank canvas widgets. If you started drawing on one widget on your desktop, you could not start drawing on the other widget using your laptop or iPad until the desktop's mouse was released. If you start another drawing stroke with the desktop while the second canvas is still blank, you don't have to wait for the desktop to finish before starting on the second canvas with either the laptop or iPad?

**Comment by 96LawDawg** [comment]
> So you start with one room with two blank canvas widgets. If you started drawing on one widget on your desktop, you could not start drawing on the other widget using your laptop or iPad until the desktop's mouse was released. If you start another drawing stroke with the desktop while the second canvas is still blank, you don't have to wait for the desktop to finish before starting on the second canvas with either the laptop or iPad?

Almost correct.  The laptop or iPad canvas has to have something on it.  If I started another drawing stroke with the desktop while the laptop/iPad was still empty, the lock-out continued.

**Comment by 96LawDawg** [comment]
The pastel palette works.  Maybe the others do also, but the changes are just too subtle to notice?  I took a shot at adding the canvas widget to the wiki at https://github.com/ArnoldSmith86/virtualtabletop/wiki/Widget-Properties.

**Comment by RaphaelAlvez** [comment]
> The pastel palette works. Maybe the others do also, but the changes are just too subtle to notice? I took a shot at adding the canvas widget to the wiki at https://github.com/ArnoldSmith86/virtualtabletop/wiki/Widget-Properties.

did you try to change the active color? some pallets use the same base colors (two first colors in color map)

**Review by 96LawDawg — APPROVED** [review]
I checked in the "colorMap" array in the main widget and the colors do all change when you select the different palettes.  I'm still not sure what "gray scale" means in this context because there are soft pinks and blues.  But whatever.  This generally works as expected.  Approved.

**Comment by RaphaelAlvez** [comment]
> I checked in the "colorMap" array in the main widget and the colors do all change when you select the different palettes.  I'm still not sure what "gray scale" means in this context because there are soft pinks and blues.  But whatever.  This generally works as expected.  Approved.

Yeah it seems to be broken. I was sure I had fixed it.

**Comment by 96LawDawg** [comment]
I tested canvas widgets in the staging server #561.  The problem with iPad and computer use at the same time was resolved there.  Also, simultaneous drawing on the same canvas worked.  The only issue I see outstanding is the gray palette.

**Comment by robartsd** [comment]
Fixed grey scale option. Iteration stopped before checking if grey was selected. I attempted to turn it into some sort of for each, but failed so took the easy route and changed the number.

**Comment by 96LawDawg** [comment]
Tested gray scale.  There are no other issues.  Recommend merging.

**Pull request merged**